### DISCUSSION TITLE
issue: Controlled inputs unregister out of order

### DISCUSSION THREAD

**Discussion by markjaniczak** [react-hook-form/react-hook-form#10397] [Q&A]
### Version Number

7.43.9

### Codesandbox/Expo snack

https://codesandbox.io/s/jolly-spence-nmzvqf?file=/src/ControlledFileInput.tsx

### Steps to reproduce

1. Go to CS
2. Click on dropdown for 'Type'
3. Select 'URL'
4. Click on dropdown for 'Type'
5. Select 'File'

### Expected behaviour

The idea here is to conditionally render the input component for a field in my form using the value from within the form. One input component accepts an array of files whereas the other accepts a URL string.

I expect that when the value for `type` changes, the current input component unmounts and the new input component mounts afterwards. Because I am providing a default value to both types of components, I would expect that the current input component resets the default value before the next one renders so it should be safe.

I have tried a couple of things:
- I've used `shouldUnregister` expecting that as the the old input component unmounts it resets the value before the new component is rendered
- I've used `watch` to check the type and trigger a `useEffect` to manually reset the value

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Edge

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by markjaniczak** [comment]
I managed to work it out using the callback form of watch and manually unregistering `value` when `type` changed:

[https://codesandbox.io/s/charming-dan-1ue9dw?file=/src/App.tsx](https://codesandbox.io/s/charming-dan-1ue9dw?file=/src/App.tsx)

Would this be the recommended approach to fixing this issue?

**Comment by Moshyfawn** [comment]
To reset a field value, try [the `resetField` method](https://react-hook-form.com/api/useform/resetfield/). unmounting or unregistering a field doesn't reset it's value.

**Comment by markjaniczak** [comment]
The documentation suggests unregistering does reset the value as keeping state and values are opt-in:

[https://react-hook-form.com/api/useform/unregister/](https://react-hook-form.com/api/useform/unregister/)

In any case, the solution above worked for me.
